ui/src/components/SignInDialog/index.jsx

```diff
@@ -28,18 +29,43 @@ export default class SignInDialog extends Component {
     credentialsDialogOpen: false,
   };
 
+  shouldStartOauth2LoginFlow() {
+    const query = parse(window.location.search.slice(1));
+
+    if (
+      query.client_id &&
+      query.response_type &&
+      query.scope &&
+      query.redirect_uri
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
   componentDidMount() {
-    const { onAuthorize, onClose } = this.props;
+    const {
+      shouldStartOauth2LoginFlow,
+      props: { onAuthorize, onClose },
+    } = this;
 
     window.addEventListener(
       'message',
-      function handler(e) {
+      async function handler(e) {
```

Comment:
does addEventListener support async functions?

Reply:
I believe so. I think async just doesn't work in constructors.

Reply:
Hm, [addEventListener docs](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#The_event_listener_callback) dont't say so, and the [handleEvent](https://developer.mozilla.org/en-US/docs/Web/API/EventListener/handleEvent#Return_value) docs also suggest the opposite.  Anyway, this is a minor issue -- a `.catch(..)` should fix it up.

Reply:
I just tested on the branch master via:

```diff
diff --git a/ui/src/components/SignInDialog/index.jsx b/ui/src/components/SignInDialog/index.jsx
index abe48e58d..eddd0f7de 100644
--- a/ui/src/components/SignInDialog/index.jsx
+++ b/ui/src/components/SignInDialog/index.jsx
@@ -28,16 +28,26 @@ export default class SignInDialog extends Component {
     credentialsDialogOpen: false,
   };
 
+  sleep = () =>
+    new Promise(resolve => {
+      window.setTimeout(resolve, 3000);
+    });
+
   componentDidMount() {
     const { onAuthorize, onClose } = this.props;
+    const { sleep } = this;
 
     window.addEventListener(
       'message',
-      function handler(e) {
+      async function handler(e) {
         if (e.origin !== window.origin || !e.data || e.data.type !== 'login') {
           return;
         }
 
+        console.time('sleep');
+        await sleep();
+        console.timeEnd('sleep');
+
         onAuthorize(UserSession.create(e.data));
         window.removeEventListener('message', handler);
         onClose();
```

The console shows `sleep: 3005ms - timer ended`

Reply:
I think the docs just say that  it won't consider the return value but there's nothing stopping the callback from being asynchronous.

Reply:
The usual concern is around error handling -- in Node an async error would be considered an UnhandledException and that's a big no-no, but it looks like the browser reports that error the same way it would a synchronous error in an event handler.  Sorry for the noise :)

Reply:
No worries :)